docs/proposals/cross-tenant-query-federation.md

```diff
@@ -0,0 +1,108 @@
+#### Challenge
+
+The tenant ID is currently read from the `X-Scope-OrgID` HTTP header. The tenant ID has those [documented limitations][cortex_tenant_id] of values being allowed.
+
+[cortex_tenant_id]: https://cortexmetrics.io/docs/guides/limitations/#tenant-id-naming
+
+#### Proposal
+
+For the query path a user should be able to specify a `X-Scope-OrgID` header with multiple tenant IDs.  Multiple tenant IDs should then be propagating throughout out the system until it reaches the querier. The `Queryable` instance returned by the querier component, is wrapped by a `mergeQueryable`, which will aggregate the results from a `Queryable` per tenant and hence treated by the downstream components as a single tenant query.
+
+To allow such queries to be processed we suggest that an experimental configuration flag `-querier.tenant-federation.enabled` will be added, which is switched off by default. Once enabled the value of the `X-Scope-OrgID` header should be interpreted as `|` separated list of tenant ids. Components which are not expecting multiple tenant ids (e.g. the ingress path) must signal an error if multiple are used.
+
+
+### Implementing Limits, Fairness and Observability for Cross-Tenant queries
+
+#### Challenge
+
+In Cortex the tenant id is used as the primary identifier for those components:
+
+- The limits that apply to a certain query.
+
+- The query-frontend maintains a per tenant query queue to implement fairness.
+
+- Relevant metrics about the query are exposed under a `user=` label.
+
+Having a query spanning multiple tenants, the existing methods are no longer correct.
+
+#### Proposal
+
+The identifier for aforementioned features for queries involving more than a single tenant should be derived from: An ordered, distinct list of tenant IDs, which is joined by a `|`.
+
+While this feature is considered experimental, this provides some insights and ability to limit multi-tenant queries with these short comings:
+
+- Cardinality costs to the possible amount of tenant ID combinations.
+
+- Query limits applied to single tenants part of a multi tenant query are ignored.
+
+
+## Conclusion
+
+| Challenge                                                                | Status                       |
+|--------------------------------------------------------------------------|------------------------------|
+| Aggregate data without overlap                                           | Implementation in PR [#3250] |
+| Exposure of feature to the user                                          | Implementation in PR [#3250] |
+| Implementing Limits, Fairness and Observability for Cross-Tenant queries | Implementation in PR [#3250] |
+
+[#3250]: https://github.com/cortexproject/cortex/pull/3250
+
+### Future work
+
+Those features are currently out of scope for this proposal, but we can foresee some interest implementing them after this proposal.
+
+#### Cross Tenant support for the ruler
+
+Ability to use multi tenant queries in the ruler.
```

Comment:
Rule groups are user specific. How do you imagine for cross-tenant rules to work?

Reply:
Fair enough, my bad.

Reply:
> I can see that eventually a concept for a User identity across Cortex, with a User having access to multiple tenants

Funnily enough, this is basically what I suggested in another comment.